src/stream/src/from_proto/source/trad_source.rs

```diff
@@ -50,8 +52,31 @@ impl ExecutorBuilder for SourceExecutorBuilder {
                 let source_name = source.source_name.clone();
                 let source_info = source.get_info()?;
 
+                let mut source_columns = source.columns.clone();
+
+                {
+                    // compatible code: introduced in https://github.com/risingwavelabs/risingwave/pull/13707
+                    // for upsert and (avro | protobuf) overwrite the `_rw_key` column's ColumnDesc.additional_column_type to Key
+                    if source_info.format() == FormatType::Upsert
+                        && (source_info.row_encode() == PbEncodeType::Avro
+                            || source_info.row_encode() == PbEncodeType::Protobuf)
+                    {
+                        let _ = source_columns.iter_mut().map(|c| {
+                            let _ = c.column_desc.as_mut().map(|desc| {
+                                let is_bytea = desc
+                                    .get_column_type()
+                                    .map(|col_type| col_type.type_name == PbTypeName::Bytea as i32)
+                                    .unwrap();
+                                if desc.name == DEFAULT_KEY_COLUMN_NAME && is_bytea {
+                                    desc.additional_column_type = AdditionalColumnType::Key as i32;
+                                }
+                            });
+                        });
+                    }
+                }
```

Comment:
Can you add a plan version in proto like this and only do the compatible logic for the old plan? Because we might change the `DEFAULT_KEY_COLUMN_NAME` later https://github.com/risingwavelabs/risingwave/blob/ef6d48316887ed771f0d9c43dba58b520b23aeba/proto/stream_plan.proto#L301-L312

Reply:
maybe a const function converting the proto Version to the DEFAULT_KEY_COLUMN_NAME?

Reply:
added a version enum in column desc.

Reply:
https://github.com/risingwavelabs/risingwave/blob/412e4bcaba05f78125ae0c36596df0419d9e6a97/src/stream/src/from_proto/source/trad_source.rs#L72-L77 checks both versioned column name and the column version itself